libcxx/include/__numeric/ranges_iota.h

```diff
@@ -0,0 +1,54 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___NUMERIC_RANGES_IOTA_H
+#define _LIBCPP___NUMERIC_RANGES_IOTA_H
+
+#include <__algorithm/out_value_result.h>
+#include <__config>
+#include <__ranges/concepts.h>
+#include <__utility/as_const.h>
+
+#if !defined(_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER)
+#  pragma GCC system_header
+#endif
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+#if _LIBCPP_STD_VER >= 23
+namespace ranges {
+template <typename _Out, typename _Tp>
+using iota_result = ranges::out_value_result<_Out, _Tp>;
+
+struct __iota_fn {
+  template <input_or_output_iterator _Out, sentinel_for<_Out> _Sent, weakly_incrementable _Tp>
+    requires indirectly_writable<_Out, const _Tp&>
+  constexpr iota_result<_Out, _Tp> operator()(_Out __first, _Sent __last, _Tp __value) const {
```

Comment:
```suggestion
  static constexpr iota_result<_Out, _Tp> operator()(_Out __first, _Sent __last, _Tp __value) {
```
same below

Reply:
@philnik777, I'm happy to make these changes, but could you explain why `static` is more appropriate than `const` here?

As an alternative, after I've added the helper function, I could make that static and leave the public functions as `const` like it says in the original paper. I stumbled on this strategy (it's used by `ranges::for_each` and `ranges::generate` at least) while looking at how other libcxx helper functions were structured.

Reply:
`static operator()` has been introduced in C++23, so we can't (currently) use it for C++20 algorithms. Otherwise we would also use it there. The simple reason to use it is that we don't care about the object and the `static` version has slightly better code gen (https://godbolt.org/z/8E1f77q6f).

Reply:
Ah I see, I wasn't aware that `static operator()` was introduced in C++23. That clears things up, thanks for the example.